### ISSUE TITLE
Breaks when building in release under beta & nightly.

### ISSUE THREAD

**Issue by ithinuel** [sammhicks/nostd_async#1]
When built with `--release` the check of the following loop is never re-evaluated:
https://github.com/sammhicks/nostd_async/blob/6f170891e7f4d4f6772974bf4c02007a23b9ef26/src/task.rs#L79-L83

I suspect the compiler does not expect this to be aliased and generates a branch instruction that targets the call to `run_once`.

Code generated by the compiler:
```
1000064a: 05 46         mov r5, r0
1000064c: 80 68         ldr r0, [r0, #8]
1000064e: 00 28         cmp r0, #0
10000650: 04 d0         beq 0x1000065c <nostd_async::task::JoinHandle$LT$T$GT$::join::ha06bd8d19d47152a+0x18> @ imm = #8
10000652: 2c 68         ldr r4, [r5]
10000654: 20 46         mov r0, r4
10000656: 00 f0 3e fc   bl  0x10000ed6 <nostd_async::task::Runtime::run_once::hdb24b1195a51081c> @ imm = #2172
1000065a: fb e7         b   0x10000654 <nostd_async::task::JoinHandle$LT$T$GT$::join::ha06bd8d19d47152a+0x10> @ imm = #-10
```

Reproduced with both beta and nightly's channel:
- `rustc 1.56.0-beta.3 (deef86610 2021-09-17)`
- `rustc 1.57.0-nightly (aa7aca3b9 2021-09-30)`

Stable (`rustc 1.55.0 (c8dfcfe04 2021-09-06)`) channel generates a different assembly for that piece of code:

```
10000648: 05 46         mov r5, r0
1000064a: 80 68         ldr r0, [r0, #8]
1000064c: 00 28         cmp r0, #0
1000064e: 05 d0         beq #10 <nostd_async::task::JoinHandle$LT$T$GT$::join::h25e6d5569edafbe4+0x1c>
10000650: 28 68         ldr r0, [r5]
10000652: 00 f0 38 fc   bl  #2160
10000656: a8 68         ldr r0, [r5, #8]
10000658: 00 28         cmp r0, #0
1000065a: f9 d1         bne #-14 <nostd_async::task::JoinHandle$LT$T$GT$::join::h25e6d5569edafbe4+0x10>
```

**Comment by sammhicks** [comment]
I've found a soundness bug which should be the cause of the problem.
The fix is in the branch [improved_soundness](https://github.com/sammhicks/nostd_async/tree/improved_soundness).
My tests pass, but if you could check that you're happy, I'll merge and publish the changes.

**Comment by ithinuel** [comment]
I confirm this branch solves the issue :)

**Comment by sammhicks** [comment]
Closed by 65733350cf1573311bb781a13b4f98f0828596a9
